sdk/cosmos/azure-cosmos/azure/cosmos/_global_partition_endpoint_manager_per_partition_automatic_failover.py

```diff
@@ -0,0 +1,247 @@
+        :return: True if the move was successful, False otherwise.
+        :rtype: bool
+        """
+        with self._lock:
+            if endpoint_region != self.current_region:
+                regional_endpoint = available_account_regional_endpoints[self.current_region]
+                request.route_to_location(regional_endpoint)
+                return True
+
+            for regional_endpoint in available_account_regional_endpoints:
+                if regional_endpoint == self.current_region:
+                    continue
+
+                if regional_endpoint in self.unavailable_regional_endpoints:
+                    continue
+
+                self.current_region = regional_endpoint
+                logger.warning("PPAF - Moving to next available regional endpoint: %s", self.current_region)
+                regional_endpoint = available_account_regional_endpoints[self.current_region]
+                request.route_to_location(regional_endpoint)
+                return True
+
+            return False
+
+class _GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover(_GlobalPartitionEndpointManagerForCircuitBreaker):
+    """
+    This internal class implements the logic for partition endpoint management for
+    geo-replicated database accounts.
+    """
+    def __init__(self, client: "CosmosClientConnection"):
+        super(_GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover, self).__init__(client)
+        self.partition_range_to_failover_info: Dict[PartitionKeyRangeWrapper, PartitionLevelFailoverInfo] = {}
+        self.ppaf_thresholds_tracker = _PPAFPartitionThresholdsTracker()
+
+    def is_per_partition_automatic_failover_enabled(self) -> bool:
+        if not self._database_account_cache or not self._database_account_cache._EnablePerPartitionFailoverBehavior:
+            return False
+        return True
+
+    def is_per_partition_automatic_failover_applicable(self, request: RequestObject) -> bool:
+        if not self.is_per_partition_automatic_failover_enabled():
+            return False
+
+        if not request:
+            return False
+
+        if (self.location_cache.can_use_multiple_write_locations_for_request(request)
+                or _OperationType.IsReadOnlyOperation(request.operation_type)):
+            return False
+
+        # if we have at most one region available in the account, we cannot do per partition automatic failover
+        available_regions = self.compute_available_preferred_regions(request)
+        if len(available_regions) <= 1:
+            return False
+
+        # if the request is not a non-query plan document request
+        # or if the request is not executing a stored procedure, return False
+        if (request.resource_type != ResourceType.Document and
+                request.operation_type != _OperationType.ExecuteJavaScript):
+            return False
+
+        return True
+
+    def try_ppaf_failover_threshold(
```

Comment:
Q: Does `try_ppaf_failover_threshold` need to happen under a lock as this has state updates?

Reply:
`ppaf_thresholds_tracker` does get updated in a sequential manner - so yes, we are probably missing a lock here, thanks for the callout! just added it